src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets

```diff
@@ -113,6 +121,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       <Reference Include="%(_NETStandardLibraryNETFrameworkLib.FileName)">
         <HintPath>%(_NETStandardLibraryNETFrameworkLib.Identity)</HintPath>
         <Private>false</Private>
+        <FindDependencies>$(FindNETFrameworkExtensionDependenciesInResolveAssemblyReferences)</FindDependencies>
```

Comment:
There's one piece of code in RAR that I'm worried about with this change https://github.com/Microsoft/msbuild/blob/cb6c2643fe4a072f28e685bc46df3bace17a7108/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs#L2305-L2306

Reply:
Ya, that's worrisome. :(

Reply:
So, under what circumstance would System.Runtime or netstandard not be found? The search is being done on [dependencyTable.Reference.Keys](https://github.com/Microsoft/msbuild/blob/cb6c2643fe4a072f28e685bc46df3bace17a7108/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs#L2258), which contains the flat list of all references. I assume the nuget graph includes those two, right? `_ComputeLockFileReferences` seems to add both for a netcoreapp.

Reply:
It's definitely broken (and thankfully we have tests that catch it).

The problem is when netfx project pulls in refs from nuget that depend on System.Runtime, but do not pull in System.Runtime from nuget. The expectation is that RAR will still detect the reference to System.Runtime and then ImplicitlyExpandDesignTimeFacades will use that detection to add TargetFrameworkDir\Facades\*.dll to ReferencePath.

Here is a minimal repro that is broken with the current iteration:

``` xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net46</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="System.Collections.Immutable" Version="1.4.0" />
  </ItemGroup>
</Project>
```

``` C#
class P {
    static void Main() {
        System.Collections.Immutable.ImmutableList<string>
            .Empty
            .Add("hello");
    }
}
```

```
Program.cs(5,14): error CS0012: The type 'Object' is defined in an assembly that is not referenced. 
You must add a reference to assembly 'System.Runtime, Version=4.0.0.0, Culture=neutral, 
PublicKeyToken=b03f5f7f11d50a3a'. [D:\Temp\repro3\repro3.csproj]
```

Reply:
Mkay, so the assumption that Nuget is responsible with resolving the closure of dependencies is actually valid only when the nupkg dependency graph matches the underlying assembly dependency graph (for each assembly edge there is a nupkg edge, or something like that). When is this guaranteed? Is it always guaranteed for netcore and netstandard?

Either way, this means that metadata to skip dependencies should only be added when the nuget graph is guaranteed to match the assembly graph. No idea when this holds. Only on specific tfms? Maybe the tfm compatibility matrix should also store information whether the tfm depends on facades.

Does this mean we can't do the optimization?

Reply:
I think we can handle facade injection as a special case. I have some work in progress on it where we use the same technique as design-time builds (still search primary references). It is the one case where RAR (actually RAR + ImplicitlyExpandDesignTimeFacades) actually add new items to ReferencePath. If there are other such cases (I'm not aware of any), they also invaldiate the assumption...

This case is actually not as bad as AutoGenerateBindingRedirects where I'm finding that we do need to do the entire dependency walk (but can skip search for .pdb, .xml, serialization assemblies).

Reply:
It's not quite that rigid @cdmihai I think we just need to also set the UsingSystemRuntime or UsingNetstandard as a result certain packages being referenced in the assets file. @nguerrera is that what you were thinking?

> This case is actually not as bad as AutoGenerateBindingRedirects where I'm finding that we do need to do the entire dependency walk (but can skip search for .pdb, .xml, serialization assemblies).

Ugh, wuttttt. Why do we need to expand the graph for binding redirects? Are we missing assemblies?

Reply:
Think of a unified assembly graph having edges labeled with the version of the assembly reference. We're not missing any nodes in that graph, but we're missing edges.

Binding redirects are needed when there are incoming edges at multiple versions.

We could consider brute force redirecting every reference or discovering the edges more cheaply than the current dependency walk: we only need to enumerate references but not probe for them. I'm inclined to tackle that in a second change. The approach for now is to leave the dependency walk enabled when generating binding redirects. 

As for detecting UsingSystemRuntime/UsingNetStandard from packages in the graph, I'm not sure that's possible in the general case. Today, it would work to put an assembly referencing system.runtime in a net45 package folder and not have a package dependency on anything. 

So what I'm doing instead is to use the same fallback as the design time build to still do a shallow search. We can maybe do something better but again I want to proceed in steps.

Reply:
To summarize: RAR extracts data necessary for binding redirection from the closure. If nugets are skipped, then it does not have that data for nuget's closure.

- it would be nice if binding redirection were a separate concern, done by another task.  Worst case it would be a lot slower, because it will have to do the closure walk again. Best case, for .netcore, it wouldn't be that much slower because much of the closure walk would be skipped in RAR. Only the walks from p2ps and non-nuget references would be redone. Or maybe msbuild could provide some shared memory model for tasks, so they can reuse in-memory caches and whatnot.
- 
- could nuget provide this data as even yet more metadata? For each reference, the metadata could be a list of all incoming version edges. Then, RAR can unify that with whatever it finds from the non-nuget closures. But I guess this is impossible because the nuget graph is not the same as the assembly graph. At best they are the same only by convention / accident :(

It would have been really nice if nuget had a different architecture, where each language could implement its own nuget graph extension of what it means to be a valid graph, and what conflict resolution means. Then, the integration with RAR could have been more tight.

Reply:
Binding redirect computation doesn't require a graph walk at all though. I get that RAR might use that information but it just requires the direct references (the edges) and all of the selected versions. The shallow check that design time builds do for netstandard and System.runtime might be enough to do this.

Reply:
Current status: 
1. If auto-generate-binding redirects: don't skip find dependencies, but do skip find .xml, .pdb, serialization/satellites assembly.
2. Borrow from design-time code to fix facade injection approach

I think we can do better than this, but I'd like to focus on testing from this point on for this change.